packages/algorand-js/src/runtime/runtime.ts

```diff
@@ -222,11 +194,14 @@ export class Runtime {
     this.createApplicationTx(flags, payFlags);
     await this.run(program); // execute TEAL code with appId = 0
 
-    this.store.globalApps.set(++this.appCounter, app.params); // update globalApps Map
-    this.ctx.state.globalApps.delete(0); // remove zero app
+    // create new application in globalApps map
+    this.store.globalApps.set(++this.appCounter, app.params);
+
+    this.ctx.state.globalApps.delete(0); // remove zero app from context after execution
+    senderAcc.createdApps.delete(0); // remove zero app from sender's account
 
-    // update local state
-    senderAcc.createdApps.pop();
+    // set new application in sender's account
+    // after setting in globalApps
     senderAcc.addApp(this.appCounter, flags);
     this.store.accounts.set(sender.addr, senderAcc);
```

Comment:
please add this in deleteApp and optInToapp too, as we are doing same thing there. 
also please add TO-DO in these functions.

Reply:
Ok, will do. 
Also this function needs to be updated also as per our discussion, line `205` will always create a fresh app with new state, not the update one. Adding TODO's and task link

Reply:
> please add this in deleteApp and optInToapp too, as we are doing same thing there.

I checked delete and opt-in functions, delete is fine, as we are just deleting the app from account and global Map. 
In optIn we need to update `globaApps` map as well, but since we will be removing this, i have added a TODO there.